src/storage-client/src/source/persist_source.rs

```diff
@@ -238,20 +253,59 @@ where
         });
 
         let mut done = false;
-        while !done {
+        while !done && !pending_parts.is_empty() {
             while let Ok(leased_part) = consumed_part_rx.try_recv() {
                 subscription
                     .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
             }
 
-            let (parts, progress) = subscription.next().await;
-            // If `until.less_equal(progress)`, it means that all subsequent batches will
-            // contain only times greater or equal to `until`, which means they can be dropped
-            // in their entirety. The current batch must be emitted, but we can stop afterwards.
-            if timely::PartialOrder::less_equal(&until_clone, &progress) {
-                done = true;
+            // if the frontier of the dataflow has advanced, subtract the
+            // memory we set aside for their corresponding in-flight parts
+            let frontier = probe.with_frontier(|frontier| frontier.to_owned());
+            while let Some((progress, size_in_bytes)) = inflight_parts.front() {
+                if PartialOrder::less_equal(progress, &frontier) {
+                    let (progress, size_in_bytes) = inflight_parts.pop_front();
+                    inflight_bytes -= size_in_bytes;
+                    info!(
+                        "{}: restoring {} bytes. now: {}",
+                        data_shard, size_in_bytes, inflight_bytes
+                    );
+                }
+            }
+
+            while let Some((progress, parts, size_in_bytes)) = pending_parts.front() {
+                if inflight_bytes + size_in_bytes < max_inflight_bytes {
+                    info!(
+                        "{}: putting {} bytes in-flight. now: {}",
+                        data_shard,
+                        size_in_bytes,
+                        inflight_bytes + size_in_bytes
+                    );
+                    inflight_bytes += size_in_bytes;
+                    let (progress, parts, size_in_bytes) =
+                        pending_parts.pop_front().expect("front was Some");
+                    inflight_parts.push_back((progress.clone(), size_in_bytes));
+                    yield (parts, progress);
+                } else {
+                    // TODO: yield until probe has frontier update?
```

Comment:
@petrosagg any ideas on how this might look? the thinking is we take in a Probe from the last stage of the dataflow to feed back in here, and we should activate (ideally `await` maybe?) this operator again once the Probe frontier advances

Reply:
I think it would be much easier if you didn't put this logic in this async_stream construction. The stream is currently set up to produce `(parts, progress)` pairs and I think we need to change to produce `(parts, size, progress)` triplets which basically communicate how many bytes are in this particular slices of times.

Then if you take the feedback stream input from my flow control PR as the mechanism by which the output frontier is communicated you can alter the main logic function like so:

```rust
let mut inflight_bytes = 0;
let mut inflight_parts: Vec<(Antichain<Timestamp>, usize)> = Vec::new();

|frontiers| {
    let output_upper = frontiers[0].frontier();
    // Retire completed slices
    let retired_slices = inflight_parts.drain_filter_swapping(|(upper, _size)| {
        PartialOrder::less_equal(slice_upper, output_upper)
    });
    for (_upper, size) in retired_slices {
        inflight_bytes -= size;
    }

    // Produce data if we should
    while inflight_bytes < max_inflight_bytes {
        if let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
            match item {
                Some(Ok((parts, size, upper))) => {
                    inflight_parts.push((upper, size));
                    inflight_bytes += size;
                        
                    // rest of the current logic
                }
                ...
            }
        }
    }
}
```

Reply:
ooh that makes a ton of sense, thank you 🙇